DDInstall/DDInstall.psm1

```diff
@@ -0,0 +1,215 @@
+
+    .OUTPUTS
+    None.
+
+    .EXAMPLE
+    C:\PS> Install-DDAgent -ApiKey "<YOUR_DATADOG_API_KEY>"
+
+    .EXAMPLE
+    C:\PS> Install-DDAgent -ApiKey "<YOUR_DATADOG_API_KEY>" -WithAPMTracers "DotNet"
+
+    .LINK
+    Learn more about the Datadog Windows Agent User:
+    https://docs.datadoghq.com/agent/guide/windows-agent-ddagent-user/
+    #>
+}
+
+###############################
+# Unexported helper functions #
+###############################
+
+function formatAgentInstallerParameters($params)
+{
+    $formattedparams = "APIKEY=$($params.ApiKey)"
+
+    if ($params.ContainsKey('Site'))                        { $formattedparams += " SITE=$($params.Site)"}
+    if ($params.ContainsKey('Tags'))                        { $formattedparams += " TAGS=``$($params.Tags)``" }
+    if ($params.ContainsKey('Hostname'))                    { $formattedparams += " HOSTNAME=$($params.Hostname)" }
+    if ($params.ContainsKey('DDAgentUsername'))             { $formattedparams += " DDAGENTUSER_NAME=$($params.DDAgentUsername)" }
+    if ($params.ContainsKey('DDAgentPassword'))             { $formattedparams += " DDAGENTUSER_PASSWORD=$($params.DDAgentPassword)" }
+    if ($params.ContainsKey('ApplicationDataDirectory'))    { $formattedparams += " APPLICATIONDATADIRECTORY=$($params.ApplicationDataDirectory)" }
+    if ($params.ContainsKey('ProjectLocation'))             { $formattedparams += " PROJECTLOCATION=$($params.ProjectLocation)" }
+
+    return $formattedparams
+}
+
+function downloadAsset($url, $outFile)
+{
+    # Suppress progress bar to prevent performance issues (affects older windows versions, see https://github.com/PowerShell/PowerShell/issues/13414)
+    $ProgressPreference = 'SilentlyContinue'
+
+    Invoke-WebRequest -UseBasicParsing -Uri $url -OutFile $outFile
+    if (-not $?)
+    {
+        $exception = [Exception]::new("Download failed.")
+        $PSCmdlet.ThrowTerminatingError([System.Management.Automation.ErrorRecord]::new($exception, "FatalError", [Management.Automation.ErrorCategory]::InvalidOperation, $null))
+    }
+
+    if (-not (hasValidDDSignature -asset $outFile))
+    {
+        $exception = [Exception]::new("$outFile did not pass the signature check.")
+        $PSCmdlet.ThrowTerminatingError([System.Management.Automation.ErrorRecord]::new($exception, "FatalError", [Management.Automation.ErrorCategory]::InvalidOperation, $null))
+    }
+}
+
+function hasValidDDSignature($asset)
+{
+    if (-not (Test-Path $asset -PathType Leaf)) {
+        $exception = [Exception]::new("$asset not found")
+        $PSCmdlet.ThrowTerminatingError([System.Management.Automation.ErrorRecord]::new($exception, "FatalError", [Management.Automation.ErrorCategory]::InvalidOperation, $null))
+    }
+
+    $signature = Get-AuthenticodeSignature $asset
+
+    if ($signature.Status -ne "Valid")
+    {
+        return $false
+    }
+
+    if ($signature.SignerCertificate.Subject -ne 'CN="Datadog, Inc", O="Datadog, Inc", L=New York, S=New York, C=US')
```

Comment:
Not sure how feasible this is, but same comment here.  If the next code signing cert someone puts "Datadog, Inc." in the CSR, then I think this will break.